readme.md
# sadcat

My ssh config is hell. Even if its not much, its hell. I have ~190 host
entries which results in 1094(!) lines of sshconfig. Which I maintain
manually. Man-u-a-ll-y!

Why is that?

I use a lot of aliases to remember all my hosts. Which is probably my fault
but im used to it. A typical entry looked like

```
Host nyc-cexapsdrap21.company.com drap21 pdrap21
  Hostname nyc-cexapsdrap21.company.com
  User myuser
  Port 22
  IdentiyFile ~/.ssh/project_id_rsa
```

You can do a lot of stuff with wildcards in sshconfig. What you cant do is
having dynamic aliases (at least what I know). This would require
a templating like option.

And thats why i wrote this tiny script.

### install & usage

as usual.

    git clone https://github.com/noqqe/sadcat
    pip install -r requirements.txt
    ./sadcat conf.toml

### config

#### ranges

A minimal config would look like this.

``` toml
[hosts]

[hosts.nyc-dpzzt]
hostname = "nyc-dpzzt[01-03]"
```

and `sadcat` generates 3 ssh entries for you called `nyc-dpzzt01`,
`nyc-dpzzt02` and `nyc-dpzzt03`. Simple.

#### templates

To save more lines you can apply a template to a hosts group

``` toml
[hosts]

[hosts.twoleadingzeros]
hostname = "fra1024mfoo[001-005]"
template = "fra1024"

[templates]

[templates.fra1024]
user = "otheruser"
port = "2202"
```

Of course, setting a variable in `hosts` will always overwrite those being
set in `templates`.

There is a special template that is applied to every host if its configured
called "default". (`[templates.default]`)

### aliases

those can be specified in `hosts` using a `toml` array if multiple or
a string. Range numbers (if available) will be applied at the end. Thats
just how i like it. Theres no deeper meaning.

=======
```
[hosts.hostsaliases]
hostname = "nyc-dpzzt[5-9]"
template = "company"
alias = [ "dypppt", "dzzzpt" ]
```

would result in

```
Host nyc-dpzzt5 dypppt5 dzzzpt5
  hostname ...

Host nyc-dpzzt6 dypppt6 dzzzpt6
  hostname ...
```

but as said, `alias` can also be just a string

```
[hosts.hostsaliases]
hostname = "nyc-dpzzt[5-9]"
alias = "dypppt"
```

#### single hosts

As you might expected, if you dont have a `Range` defined in a hostname
this entry will result in one single host.

``` toml
[hosts.singlehost]
hostname = "fra1024mfoo23"
user = "foo"
```

### custom

If all that is still not enough flexibility, you can add custom snippets to
your ssh config by defining multiline strings in the `[custom]` section
with full hosts.

``` toml
[custom]

hostsb = '''

Host bar.foo
  hostname bla
  user foo

Host foo.bar
  hostname foo
  user bar
  port par
'''

strangehost = '''

Host baz
  hostname baz
  port 666
  user evil
'''
```

For more detailed examples see `conf.toml` in this repo.

### sadcat?

I used a project name generator and liked it.


HISTORY.rst
Changelog
=========


v3.0.0 (2019-10-10)
-------------------
- Release: 3.0.0. [Baumann]
- Release fix. [Baumann]
- Changelog: 2.0.0. [Baumann]


v2.0.0 (2019-10-10)
-------------------
- Release: 2.0.0. [Baumann]
- Hist. [Baumann]


v1.0.0 (2019-10-10)
-------------------
- Release: 1.0.0. [Baumann]
- Python3 Conversion. [Baumann]
- Changelog: 0.4.9. [Florian Baumann]


v0.4.9 (2017-10-10)
-------------------
- Release: 0.4.9. [Florian Baumann]
- Fix in logic. [Florian Baumann]
- Changelog: 0.4.8. [Florian Baumann]


v0.4.8 (2017-02-20)
-------------------
- Release: 0.4.8. [Florian Baumann]
- Fix template to default inheritance. [Florian Baumann]
- Changelog: 0.4.7. [Florian Baumann]


v0.4.7 (2017-02-16)
-------------------
- Release: 0.4.7. [Florian Baumann]
- Fix. [Florian Baumann]
- Changelog: 0.4.6. [Florian Baumann]


v0.4.6 (2017-02-16)
-------------------
- Release: 0.4.6. [Florian Baumann]
- Changelog: 0.4.5. [Florian Baumann]


v0.4.5 (2017-02-16)
-------------------
- Release: 0.4.5. [Florian Baumann]
- Fix release scripts. [Florian Baumann]
- Fix. [Florian Baumann]
- Fix release scripts. [Florian Baumann]
- Fix release scripts. [Florian Baumann]
- Fix release scripts. [Florian Baumann]
- Improved readme. [Florian Baumann]
- Changelog: 0.4.1. [Florian Baumann]


v0.4.1 (2017-02-16)
-------------------
- Release: 0.4.1. [Florian Baumann]
- Changelog: 0.4.0. [Florian Baumann]


v0.4.0 (2017-02-16)
-------------------
- Release: 0.4.0. [Florian Baumann]
- Changelog: 0.3.0. [Florian Baumann]


v0.3.0 (2017-02-16)
-------------------
- Release: 0.3.0. [Florian Baumann]
- Changelogfix. [Florian Baumann]
- Fix. [Florian Baumann]
- Fix. [Florian Baumann]


v0.2.9 (2017-02-16)
-------------------
- Release: 0.2.9. [Florian Baumann]


v0.2.8 (2017-02-16)
-------------------
- Release: 0.2.8. [Florian Baumann]


v0.2.7 (2017-02-16)
-------------------
- Release: 0.2.7. [Florian Baumann]
- Fix default template. [Florian Baumann]


v0.2.6 (2017-02-16)
-------------------
- Release: 0.2.6. [Florian Baumann]
- History. [Florian Baumann]


v0.2.5 (2017-02-16)
-------------------
- Release: 0.2.5. [Florian Baumann]
- Fix problem with defaults and template inheritance. [Florian Baumann]
- Changelog: 0.2.4. [Florian Baumann]


v0.2.4 (2017-02-13)
-------------------
- Release: 0.2.4. [Florian Baumann]
- Setup fixes. [Florian Baumann]
- Changelog: 0.2.3. [Florian Baumann]


v0.2.3 (2017-02-13)
-------------------
- Release: 0.2.3. [Florian Baumann]
- Satisfy pypi. [Florian Baumann]
- Changelog: 0.2.2. [Florian Baumann]


v0.2.2 (2017-02-13)
-------------------
- Release: 0.2.2. [Florian Baumann]
- Readme. [Florian Baumann]
- Changelog: 0.2.1. [Florian Baumann]


v0.2.1 (2017-02-13)
-------------------
- Release: 0.2.1. [Florian Baumann]
- Fix. [Florian Baumann]
- Changelog: 0.2.0. [Florian Baumann]
- Changelog: 0.2.0. [Florian Baumann]


v0.2.0 (2017-02-13)
-------------------
- Release: 0.2.0. [Florian Baumann]
- Changelog and dev requirements. [Florian Baumann]


v0.1.0 (2017-02-09)
-------------------
- Release: 0.1.0. [Florian Baumann]
- Release. [Florian Baumann]
- Fix template handling. [Florian Baumann]
- Ignore file. [Florian Baumann]


v0.0.1 (2017-02-09)
-------------------
- Pipyfy. [Florian Baumann]
- Main. [Florian Baumann]
- Readme to rst. [Florian Baumann]
- Default template. [Florian Baumann]
- Catch syntax errors in config. [Florian Baumann]
- Readme. [Florian Baumann]
- Custom section. [Florian Baumann]
- Readme. [Florian Baumann]
- Readme. [Florian Baumann]
- Req. [Florian Baumann]
- Help. [Florian Baumann]
- Alias as array or string possible. [Florian Baumann]
- Init. [Florian Baumann]




sadcat/sadcat.py
#!/usr/bin/env python3

import os
import sys
import textwrap
import re
import toml

def parse_config(conffile):
    """"
    open and read config file
    :conffile: file path as str
    :returns: dict
    """

    if not os.path.isfile(conffile):
        print("Error: config not found")
        sys.exit(1)

    try:
        with open(conffile) as conffile:
            config = toml.loads(conffile.read())
    except toml.TomlDecodeError as e:
        print("Error: Invalid config.")
        print("Error was: %s" % e)
        sys.exit(1)

    return config

def generate_range(start, end):
    """
    Convert [01-09] or [5-9] to a list() of values like
    [ '01', '02', '03' ] or so. This list is used for the Host
    section loop afterwards
    :start: str
    :end: str
    """

    # if it seems to be a single host entry,
    # skip all this
    if start is None and end is None:
        return [""]

    # convert str to int
    s = int(start)
    e = int(end)

    # counting var
    c = s

    # get a list
    nlist = []

    # loop in range
    while c <= e:

        # 001
        if start.startswith("00"):
            nlist.append("%03d" % c)
        # 01
        elif start.startswith("0"):
            nlist.append("%02d" % c)
        # 1
        else:
            nlist.append("%01d" % c)

        c = c + 1

    return nlist

def delkey_if_exists(sec, key):
    """
    Just a simple wrapper to delete stuff from
    a dictionary
    :sec: dict
    :key: str
    :returns: dict
    """
    try:
        del sec[key]
    except (KeyError, TypeError):
        pass
    return sec


def ssh_block(nlist, start, end, section, template, defaults):
    for i in nlist:

        # working copy of the template object
        # to be able to remove keys that are in host section
        templ = dict(template)
        defs = dict(defaults)

        # build hostname and delete
        hostname = "%s%s%s" % (start, i, end)

        # build aliases and delete
        alias = ""
        try:
            if isinstance(section["alias"], list):
                for x in section["alias"]:
                    alias = alias + x + str(i) + " "
            else:
                alias = section["alias"] + str(i)
        except KeyError:
            pass


        # create hostname and host section
        print("Host %s %s" % (hostname,alias))
        print(" hostname %s" % hostname)

        # delete all keys from templ and defaults that are set in host
        # this effect is used to overwrite variables from templ
        # in host
        for key in section:
            templ = delkey_if_exists(templ, key)
            defs = delkey_if_exists(defs, key)

        # delete keys in default when in template
        for key in templ:
            defs = delkey_if_exists(defs, key)

        # delete meta keys only used by sadcat itself
        section = delkey_if_exists(section, "template")
        section = delkey_if_exists(section, "hostname")

        # apply default vars
        for k, v in defs.items():
            print(" %s %s" % (k,v))

        # apply all templ vars
        for k, v in templ.items():
            print(" %s %s" % (k,v))

        # apply all left host section wars
        for k, v in section.items():
            if not k == "alias":
                print(" %s %s" % (k,v))

        print("")

def get_template_for_host(section, config):
    """
    get template dict object from host section
    """
    try:
        template = section["template"]
        try:
            template = config["templates"][template]
        except KeyError:
            print("ERROR: Template %s does not exist!" % template)
            sys.exit(1)
    except KeyError:
        template = {}
        pass

    return template

def parse_hostname_exp(hostgroup):
    """
    Split hostname from foo[01-03]bar into
    01, 02, foo, bar.
    Bascially a very naive parser of regex ranges.
    """

    f = re.findall(r'(.*)\[(\d+)-(\d+)\](.*)', hostgroup["hostname"])

    try:
        rstart = f[0][1]
        rend = f[0][2]
        hstart = f[0][0]
        hend = f[0][3]
    except:
        # Seems to be a single host
        rstart = None
        rend = None
        hstart = hostgroup["hostname"]
        hend = ""

    return rstart, rend, hstart, hend

def generate_hosts(config):

    hosts = config["hosts"]

    try:
        defaults = config["templates"]["default"]
    except KeyError:
        defaults = {}

    for hostgroup in hosts:

        # new dict for this host section
        section = config["hosts"][hostgroup]

        # Comment for ssh config
        print("# %s\n" % hostgroup)

        # parse hostname expression
        rstart, rend, hstart, hend = parse_hostname_exp(section)

        # generate range
        nlist = generate_range(rstart, rend)

        # get template referenced in this host entry
        template = get_template_for_host(section, config)


        # fill all gathered informations into block generator
        ssh_block(nlist, hstart, hend, section, template, defaults)

def paste_appended_options(config):
    """
    Pasts appended ssh config parts from
    ssh config if any configured
    :returns: bool
    """
    try:
        print("# Custom appended section\n")
        for c in config["custom"].keys():
            print("# " + c)

            # this is a dirty hack for https://github.com/uiri/toml/issues/71
            for line in config["custom"][c].split('\n'):
                if "Host " in line:
                    print(textwrap.dedent(line))
                else:
                    print("  " + textwrap.dedent(line))

    except KeyError:
        pass

    return True

def main():

    # help
    if len(sys.argv) <= 1:
        print("Usage: %s conf.toml" % sys.argv[0])
        sys.exit(1)

    config = parse_config(sys.argv[1])
    generate_hosts(config)
    paste_appended_options(config)

if __name__ == '__main__':
    main()
